src/cca/scripts/uproxy_server.ts

```diff
@@ -56,25 +37,79 @@ export class UproxyServer implements Server {
   }
 }
 
+// Name by which servers are saved to storage.
+const SERVERS_STORAGE_KEY = 'saved-servers';
+
+// Type of the object placed, in serialised form, in storage.
+interface SavedServers {
+  servers: { [id: string]: SavedServer }
+}
+
+// A server as saved to storage.
+interface SavedServer {
+  cloudTokens?: cloud_social_provider.Invite;
+}
+
+// Maintains a persisted set of servers and liases with the core.
 export class UproxyServerRepository implements ServerRepository {
-  constructor(private core: CoreConnector, private vpnDevice: VpnDevice) {
-    this.core.login({
-      network: 'Cloud',
-      loginType: uproxy_core_api.LoginType.INITIAL,
-    });
+  // core must already be logged into social networks.
+  constructor(
+    private storage: Storage,
+    private core: CoreConnector,
+    private vpnDevice: VpnDevice) { }
+
+  public getSavedServers() {
+    const servers = this.loadServers().servers;
+    return Promise.all(Object.keys(servers).map((host) => {
+      return this.notifyCoreOfServer(servers[host].cloudTokens);
```

Comment:
Consider (not in this PR) adding to the core on connect and removing on disconnect.

Maybe add a TODO?